### PR TITLE
Read ordered chunks from RecvStream

### PR DESCRIPTION
**Pull request by stammw**
Hello !

I've started trying to implement #929.

Still I'd better be adding some tests before its ready. So I'm setting this PR WIP for a first feedback.

### CODE REVIEW

File: quinn-proto/src/connection/assembler.rs
Comment (djc) [inline quinn-proto/src/connection/assembler.rs]: Is this actually necessary? I thought we prevented this scenario in the assembler.

File: quinn-proto/src/connection/streams.rs
Comment (djc) [inline quinn-proto/src/connection/streams.rs:245]: This is pretty repetitive, can we factor out a bunch of the match here (ideally in a separate commit)? Could apply to `read_unordered()` as well, ideally, but just between `read()` and `read_chunk()` would be good if that makes it hard.

File: quinn-proto/src/connection/assembler.rs
Comment (djc) [inline quinn-proto/src/connection/assembler.rs:78]: Does this actually happen? I thought the `Assembler` prevented this kind of thing.

File: quinn-proto/src/connection/streams.rs
Comment (djc) [inline quinn-proto/src/connection/streams.rs:1169]: A `match` expression would be more readable and more compact, I think?

File: quinn/src/streams.rs
Comment (djc) [inline quinn/src/streams.rs:446]: Also pretty repetitive. Can we factor these things?

File: quinn-proto/src/connection/assembler.rs
Comment (stammw) [reply quinn-proto/src/connection/assembler.rs]: It replicates the logic I found in `consume()`.

Comment (stammw) [reply quinn-proto/src/connection/assembler.rs]: Looks like `insert()` doesn't push already received data, but a consecutive insert containing the same data would leave the precedent useless chunk in (unless defragmentation happens ?).

File: quinn-proto/src/connection/assembler.rs
Comment (stammw) [reply quinn-proto/src/connection/assembler.rs:78]: Looks like insert() doesn't push already received data, but a consecutive insert containing the same data would leave the precedent useless chunk in (unless defragmentation happens ?).

Comment (djc) [reply quinn-proto/src/connection/assembler.rs:78]: Ah yeah, fair enough.

File: quinn/src/streams.rs
Comment (Ralith) [inline quinn/src/streams.rs:350]: This should be documented, including an explicit warning against relying on the chunk boundaries having any particular relationship to the sender's writes.

**Review by Ralith — CHANGES_REQUESTED** [review]
Thanks for working on this!

File: quinn-proto/src/connection/assembler.rs
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs:78]: In fact, under normal (ordered) operation `insert` makes no effort to deduplicate whatsoever unless defragmentation occurs, so this is very necessary. See also the analogous code in `read`.

File: quinn-proto/src/connection/assembler.rs
Comment (Ralith) [reply quinn-proto/src/connection/assembler.rs]: This case is needed per https://github.com/quinn-rs/quinn/pull/952#discussion_r549865148.

File: quinn/src/streams.rs
Comment (stammw) [reply quinn/src/streams.rs:446]: Would a macro be acceptable here ?

Comment (stammw) [reply quinn/src/streams.rs:388]: This one wasn't that hard to refactor with a generic `fn`, actually.

File: quinn-proto/src/connection/streams.rs
Comment (stammw) [reply quinn-proto/src/connection/streams.rs]: I got a solution I find a little complex in feb3b63, what do you think about this?

File: quinn-proto/src/connection/streams.rs
Comment (stammw) [reply quinn-proto/src/connection/streams.rs]: done

File: quinn-proto/src/connection/streams.rs
Comment (djc) [reply quinn-proto/src/connection/streams.rs]: It looks like you can just inline the `IntoReadResult` into the closure passed to `try_read()`? We already have the nested `map()` calls anyway. Or am I missing something?

File: quinn/src/streams.rs
Comment (stammw) [reply quinn/src/streams.rs]: I got this a better comment. I'm not sure I grasped everything you'd like it to say?

File: quinn/src/streams.rs
Comment (Ralith) [inline quinn/src/streams.rs:389]:
```suggestion
    /// Foundation of [`read_chunk()`]: RecvStream::read_chunk
```

File: quinn/src/streams.rs
Comment (Ralith) [inline quinn/src/streams.rs:376]:
```suggestion
    /// Read the next segment of data
```

File: quinn/src/streams.rs
Comment (Ralith) [inline quinn/src/streams.rs:384]: Tweaked wording a bit:
```suggestion
    /// Yields a segment of data beginning immediately after the last data yielded by
    /// `read` or `read_chunk`, or `None` if the stream was finished.
    ///
    /// Slightly more efficient than `read` due to not copying. Chunks boundaries
    /// do not correspond to peer writes, and hence cannot be used as framing.
```

File: quinn-proto/src/connection/mod.rs
Comment (Ralith) [inline quinn-proto/src/connection/mod.rs:872]:
```suggestion
    /// window is filled. When in-order delivery is required, the sibling `read()` or `read_chunk()`
    /// methods should be used.
```

**Review by Ralith — CHANGES_REQUESTED** [review]
Wording tweaks for the comments:

File: quinn-proto/src/connection/streams.rs
Comment (stammw) [reply quinn-proto/src/connection/streams.rs]: Ho yeah maybe ! :sweat_smile:
For the moment I get stuck because `self.add_read_credits(len as u64)` needs to mutate `self` to compute `transmit_max_data`, and it is already mutably borrowed by the closure. But I'll try a little harder in a bit.

**Comment by djc** [comment]
(Note that #961 is blocked on this, so would be nice to follow through soon...)

**Comment by Ralith** [comment]
My wording in #961 was unclear--I don't actually consider it blocked, I'm just holding off on adding additional refactoring that would make rebasing *this* PR harder than it otherwise would be.

**Comment by Matthias247** [comment]
While we are at it: I would also appreciate a vectored read API - which can retrieve more than 1 chunk at a time. Because locking a Mutex and potentially doing a task wakeup for reading just 1kB of data (due to Quic packet sizes) is fairly unefficient.

One API I've used in a project for that is:

```
async fn read(&mut self, &mut [Bytes]) -> Result<(usize, bool), Error>
```

Where the caller provides an array (initialized with dummy elements / `Bytes::new()`) - and the implementation fills as many array items as possible and returns the number of chunks read in the first return value. Second return value is for indicating end of stream.

**Comment by stammw** [comment]
I finally got the read methods factorization right (yeah the `IntoResult` thing was weird...). 

I've addesd a test for `assembler::read_chunk()` in the first commit.

All the wording changes have been addressed.

A new commit contains @Matthias247  suggestion. This is still WIP as I'm waiting for your feedback on this public API change, and it should probably be tested in the functional test module.

**Comment by djc** [comment]
Did you see #955?

**Comment by stammw** [comment]
I should probably have paid more attention, sorry... Now the `IntoResult` solution is totally removed here. I'm a little too tired to judge which one is better.

**Comment by djc** [comment]
Maybe @Ralith or @Matthias247 want to take a look which version they prefer.

File: quinn-proto/src/connection/assembler.rs
Comment (Matthias247) [inline quinn-proto/src/connection/assembler.rs:76]: Isn't `None` end of stream? i'm not sure how this function is used, but it might need to return a blocked status instead.

**Comment by Matthias247** [comment]
> Maybe @Ralith or @Matthias247 want to take a look which version they prefer.

I skimmed both commits, and I have a slight preference for this one since I found the `ReadResult/StreamReadResult/IntoReadResult` types a bit harder to follow.

I think this change currently has a good balance between code reuse and readability.

@stammw Also thanks for incorporating my feedback!

File: quinn-proto/src/connection/streams.rs
Comment (Matthias247) [inline quinn-proto/src/connection/streams.rs:1170]: You can implement this via `read_chunks:`

```rust
let mut bufs = &mut [Bytes::new()];
match self.read_chunks(bufs) {
    Ok(Some(1)) => Ok(Some(std::mem::replace(&mut bufs[0], Bytes::new()))),
    Ok(Some(_)) => unreachable!(),
    Ok(None) => Ok(None),
    Err(e) => Err(e),
}
```

Doesn't however really a buy a lot here. Might be more interesting if assembler would be doing less.

File: quinn-proto/src/connection/assembler.rs
Comment (stammw) [reply quinn-proto/src/connection/assembler.rs:76]: This method works as it's siblings: it returns `None` when there is nothing to *assemble*. Blocked status is managed at the *stream* level in `quinn-proto::connection::streams::read_chunk()`.

See the `chunks_dedup()` test at the end of this file for an example of expected behavior.

File: quinn-proto/src/connection/streams.rs
Comment (stammw) [reply quinn-proto/src/connection/streams.rs:1170]: To me there is no real advantage in factorizing in this way. Since it is different layers composition already, and doing so would add a tiny overhead without making the overall code shorter at all.

Comment (djc) [reply quinn-proto/src/connection/streams.rs:1170]: I think I agree that it doesn't seem to help here.

**Comment by stammw** [comment]
The current `read` methods factorization solution (without `IntoResult`) is actually better IMO.

I've added a test for `read_chunks()`, so this is ready to be merged if you guys like the current state of this PR.

File: quinn/src/streams.rs
Comment (djc) [inline quinn/src/streams.rs:382]:
```suggestion
    /// Slightly more efficient than `read` due to not copying. Chunk boundaries
    /// do not correspond to peer writes, and hence cannot be used as framing.
```

File: quinn/src/streams.rs
Comment (djc) [inline quinn/src/streams.rs:399]:
```suggestion
    /// Slightly more efficient than `read` due to not copying. Chunk boundaries
    /// do not correspond to peer writes, and hence cannot be used as framing.
```

**Review by djc — COMMENTED** [review]
LGTM, modulo small doc nits.

**Comment by stammw** [comment]
Should be ok now. 

Happy to contribute to quinn a bit :)

**Review by djc — APPROVED** [review]

**Pull request merged**
